openquakeplatform/openquakeplatform/updatures/app.py

```diff
@@ -0,0 +1,952 @@
+                maxpks[model]['maxpk'] = oldels[model].maxpk + 1
+
+    return maxpks
+
+
+def key_get(md, item):
+    if md.natural:
+        k = md.natural(item)
+    else:
+        k = [item['pk']]
+    return tuple(k)
+
+
+def group_objs(base):
+    group = {}
+    dyn_refs = {}
+    groupk = {}
+
+    for model in MODELS_ORDER:
+        group[model] = []
+        dyn_refs[model] = {}
+        groupk[model] = {}
+
+    for item in base:
+        model = item['model']
+        md = MODELS_DESCR[model]
+
+        # if dynamic references populate the ModelDescription with more
+        # virtual references
+        for ref_field, ref in md.refs.iteritems():
+            if isinstance(ref.model, types.FunctionType):
+                dyn_key = "__dynamic__.%s.fk" % ref.model(item)
+                if not dyn_refs[model].get(dyn_key, None):
+                    dyn_refs[model][dyn_key] = ModelRefs(ref.model(item), False)
+                item['fields'][dyn_key] = get_value(item, ref_field)
+
+        group[model].append(item)
+        k = key_get(md, item)
+        groupk[model][k] = item
+
+    for model in MODELS_ORDER:
+        MODELS_DESCR[model].refs = dict(MODELS_DESCR[model].refs.items()
+                                        + dyn_refs[model].items())
+
+    return (group, groupk)
+
+def kappend(groupk, model, item):
+    model = item['model']
+    md = MODELS_DESCR[model]
+
+    if md.natural:
+        k = tuple(md.natural(item))
+    else:
+        k = item['pk']
+    groupk[model][k] = item
+
+
+def update_pk(updates_gr, updatesk_gr, model, item, maxpks, new_pk):
+    """
+    updates_gr   grouped updates items
+    updatesk_gr  grouped updates items (dict with keys)
+    model        item model
+    item         item to be updated
+    maxpks       dict of maximum numeric pk reached
+    new_pk       new key
+    """
+    md = MODELS_DESCR[model]
+    if md.pk_natural:
+        return False
+
+    if item['pk'] == new_pk:
+        return True
+
+    # search a previous defined item with the same pk
+    for item_same_pk in updates_gr[model]:
+        if item_same_pk['pk'] == new_pk:
+            old_pk = item['pk']
+            item['pk'] = maxpks[model]['maxpk'] + 1
+            update_pk(updates_gr, updatesk_gr, model, item_same_pk, maxpks,
+                      maxpks[model]['maxpk'])
+            item['pk'] = old_pk
+            break
+    else:
+        item_same_pk = None
+
+    # update all references
+    if md.natural is None:
+        for ref_model, ref_md in MODELS_DESCR.iteritems():
+            # found each model has a refs value associated with
+            # the current item model
+            for ref_reffield, ref in ref_md.refs.iteritems():
+                if isinstance(ref.model, types.FunctionType):
+                    continue
+
+                if ref.model != model:
+                    continue
+
+                updatures.UPD_LOG.log(0, "MDREF: %s" % ref_model)
+
```

Comment:
What does it mean log level 0? It is less than debug, when the message should be logged?